azkaban-common/src/main/java/azkaban/utils/EmailMessageCreator.java

```diff
@@ -0,0 +1,42 @@
+package azkaban.utils;
+
+import java.util.Properties;
+import javax.inject.Inject;
+import javax.mail.NoSuchProviderException;
+
+public class EmailMessageCreator {
+
+  public static final int DEFAULT_SMTP_PORT = 25;
+
+  private final String mailHost;
+  private final int mailPort;
+  private final String mailUser;
+  private final String mailPassword;
+  private final String mailSender;
+  private final String tls;
+  private final boolean usesAuth;
+
+  @Inject
+  public EmailMessageCreator(final Props props) {
+    this.mailHost = props.getString("mail.host", "localhost");
+    this.mailPort = props.getInt("mail.port", DEFAULT_SMTP_PORT);
+    this.mailUser = props.getString("mail.user", "");
+    this.mailPassword = props.getString("mail.password", "");
+    this.mailSender = props.getString("mail.sender", "");
+    this.tls = props.getString("mail.tls", "false");
+    this.usesAuth = props.getBoolean("mail.useAuth", true);
+  }
+
+  public EmailMessage createMessage() {
+    final EmailMessage message = new EmailMessage(
+        this.mailHost, this.mailPort, this.mailUser, this.mailPassword, this);
+    message.setFromAddress(this.mailSender);
+    message.setTLS(this.tls);
+    message.setAuth(this.usesAuth);
+    return message;
+  }
+
+  public JavaxMailSender createSender(final Properties props) throws NoSuchProviderException {
+    return new JavaxMailSender(props);
```

Comment:
can you help me understand better of the purpose of this method? why do we need to wrap a method around JavaxMailSender(props)?

Reply:
A new sender needs to be created every time a message is created (in `EmailMessage.java`):

```java
    final JavaxMailSender sender = this.creator.createSender(props);
    final Message message = sender.createMessage();
```

If you compare this code with the previous implementation, you can see why we need to wrap the code that touches the final / static classes of `javax.mail`, if we want to replace them with mocks in unit tests:

```java
final Session session = Session.getInstance(props, null);
final Message message = new MimeMessage(session);
```

The `JavaxMailSender` just wraps the operations that would use `javax.mail.Session`.

As the class level javadoc on `JavaxMailSender` says:

> Wraps javax.mail features, mostly because Session is a final class and can't be mocked